validphys2/src/validphys/commondataparser.py

```diff
@@ -40,22 +108,828 @@
+            )
+            if self.apfelcomb.shifts is not None and self.shifts is None:
+                object.__setattr__(self, 'shifts', self.apfelcomb.shifts)
+                self.apfelcomb.shifts = None
+
+    def fktables_to_paths(self, grids_folder):
+        """Given a source for pineappl grids, constructs the lists of fktables
+        to be loaded"""
+        ret = []
+        for operand in self.FK_tables:
+            ret.append([grids_folder / f"{m}.{EXT}" for m in operand])
+        return ret
+
+    @classmethod
+    def parser(cls, yaml_file):
+        """The yaml databases in the server use "operands" instead of "FK_tables" """
+        if not yaml_file.exists():
+            raise FileNotFoundError(yaml_file)
+        meta = yaml.safe_load(yaml_file.read_text())
+        # Make sure the operations are upper-cased for compound-compatibility
+        meta["operation"] = "NULL" if meta["operation"] is None else meta["operation"].upper()
+        if "operands" in meta:
+            meta["FK_tables"] = meta.pop("operands")
+        return parse_input(meta, cls)
+
+    def __hash__(self):
+        """Included in the hash any piece of information that can change the
+        definition of the theory for a given dataset for functions using a cache"""
+        to_be_hashed = [self.operation, self.conversion_factor]
+        to_be_hashed.append(tuple([tuple(i) for i in self.FK_tables]))
+        if self.shifts is not None:
+            to_be_hashed.append(tuple(self.shifts.keys()))
+            to_be_hashed.append(tuple(self.shifts.values()))
+        return hash(tuple(to_be_hashed))
+
+
+## Theory end
+
+
+@dataclasses.dataclass(frozen=True)
+class Variant:
+    """The new commondata format allow the usage of variants
+    A variant can overwrite a number of keys, as defined by this dataclass
+    """
+
+    data_uncertainties: Optional[list[ValidPath]] = None
+    theory: Optional[TheoryMeta] = None
+    data_central: Optional[ValidPath] = None
+
+
+ValidVariants = Dict[str, Variant]
+
+
+### Kinematic data
+@dataclasses.dataclass(frozen=True)
+class ValidVariable:
+    """Defines the variables"""
+
+    label: str
+    description: str = ""
+    units: str = ""
+
+    def full_label(self):
+        if self.units:
+            return f"{self.label} ({self.units})"
+        return self.label
+
+    def apply_label(self, value):
+        """Return a string formatted as label = value (units)"""
+        tmp = f"{self.label} = {value}"
+        if self.units:
+            tmp += f" ({self.units})"
+        return tmp
+
+
+@dataclasses.dataclass(frozen=True)
+class ValidKinematics:
+    """Contains the metadata necessary to load the kinematics of the dataset.
+    The variables should be a dictionary with the key naming the variable
+    and the content complying with the ``ValidVariable`` spec.
+
+    Only the kinematics defined by the key ``kinematic_coverage`` will be loaded,
+    which must be three.
```

Comment:
Why must they be three? With this new commondata it should be possible to store any number

Reply:
https://www.youtube.com/watch?v=-IOMNUayJjI

(more seriously, expecting 3 values is hardcoded in so many places that I decided to leave it like that for now and then remove this limitation in the future)

Reply:
> expecting 3 values is hardcoded in so many places

Fair enough, can't say that surprises me